releasing/releasing/releasing.go

```diff
@@ -0,0 +1,478 @@
+
+		addWorktree(pwd, tempDir, branch)
+
+		merge(tempDir, "upstream/master")
+		// Update module path
+		mod.path = tempDir
+
+		logInfo(
+			"Releasing summary:\nDir:\t%s\nModule:\t%s %s\nBranch:\t%s\nTag:\t%s",
+			tempDir,
+			mod.name,
+			mod.version.String(),
+			branch,
+			mod.Tag(),
+		)
+
+		// Run module tests
+		output, err := mod.RunTest()
+		if err != nil {
+			logWarn(output)
+		} else if !noDryRun {
+			logInfo("Skipping push module %s. Run with --no-dry-run to push the release.", mod.name)
+		} else {
+			pushRelease(tempDir, branch, mod)
+		}
+		// Clean
+		removeTempDir()
+		pruneWorktree(pwd)
+		deleteBranch(pwd, branch)
+		logInfo("Releasing for module %s completes", mod.name)
+	},
+}
+
+var subCmds = [...]*cobra.Command{
+	listSubCmd,
+	release,
+}
+
+// === Main function ===
+
+func main() {
+	for _, cmd := range subCmds {
+		rootCmd.AddCommand(cmd)
+	}
+	rootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "verbose output")
+	release.Flags().BoolVarP(&noDryRun, "no-dry-run", "", false, "disable dry-run")
+	release.Flags().BoolVarP(&doTest, "do-test", "", false, "run module tests before releasing")
+
+	if err := rootCmd.Execute(); err != nil {
+		logFatal(err.Error())
+	}
+}
+
+func getModuleCurrentVersion(modName, path string) string {
+	mod := module{
+		name: modName,
+		path: path,
+	}
+	mod.UpdateCurrentVersion()
+	v := mod.version.String()
+	logDebug("module %s version.toString => %s", mod.name, v)
+	return v
+}
+
+// === module version struct and functions definition ===
+
+type moduleVersion struct {
+	major int
+	minor int
+	patch int
+}
+
+func (v moduleVersion) String() string {
+	return fmt.Sprintf("v%d.%d.%d", v.major, v.minor, v.patch)
+}
+
+func (v *moduleVersion) Set(major, minor, patch int) {
+	v.major = major
+	v.minor = minor
+	v.patch = patch
+}
+
+func (v *moduleVersion) FromString(vs string) {
+	versions := strings.Split(vs, ".")
+	major, err := strconv.Atoi(versions[0])
+	if err != nil {
+		logFatal(err.Error())
+	}
+	minor, err := strconv.Atoi(versions[1])
+	if err != nil {
+		logFatal(err.Error())
+	}
+	patch, err := strconv.Atoi(versions[2])
+	if err != nil {
+		logFatal(err.Error())
+	}
+	v.Set(major, minor, patch)
+}
+
+func (v *moduleVersion) Bump(t string) {
+	if t == "major" {
+		v.major++
+		v.minor = 0
+		v.patch = 0
+	} else if t == "minor" {
+		v.minor++
+		v.patch = 0
+	} else if t == "patch" {
+		v.patch++
+	} else {
+		logFatal("Invalid version type: %s", t)
+	}
+}
+
+// === module struct and functions definition ===
+
+type module struct {
+	name    string
+	path    string
+	version moduleVersion
+}
+
+func (m *module) UpdateCurrentVersion() {
+	logDebug("Getting latest tag for %s", m.name)
+	cmd := exec.Command("git", "tag", "-l")
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Dir = m.path
+	err := cmd.Run()
+	if err != nil {
+		logFatal(err.Error())
+	}
+
+	// Search for module tag
+	regString := fmt.Sprintf("(?m)^%s/v(\\d+\\.){2}\\d+$", m.name)
```

Comment:
please factor to something we can write a unit test for, e.g. 
```
NewModuleVersionSliceFromGitTags(string) [] moduleVersion
```
this can live in moduleversion.go